src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryUnsafeModifier/UnnecessaryUnsafeModifierUtilities.cs

```diff
@@ -0,0 +1,151 @@
+        CancellationToken cancellationToken)
+    {
+        var originalTree = semanticModel.SyntaxTree;
+        var originalRoot = originalTree.GetRoot(cancellationToken);
+
+        using var _1 = ArrayBuilder<SyntaxNode>.GetInstance(out var nodesToCheck);
+
+        foreach (var existingNode in originalRoot.DescendantNodes())
+        {
+            if (existingNode is not MemberDeclarationSyntax and not LocalFunctionStatementSyntax)
+                continue;
+
+            if (GetUnsafeModifier(existingNode) != default &&
+                ShouldAnalyzeNode(semanticModel, existingNode, cancellationToken))
+            {
+                nodesToCheck.Add(existingNode);
+            }
+        }
+
+        // We actually process things in three passes.  That way We can tell if containing unsafe modifiers are
+        // unnecessary, even if inner ones are necessary.  For example, consider an unsafe type with an unsafe method
+        // inside of it.  We don't want to remove both 'unsafe' modifiers and have them both be considered necessary
+        // just because the method one was actually the important one.
+
+        foreach (var group in nodesToCheck.GroupBy(node =>
+            node switch
+            {
+                LocalFunctionStatementSyntax => NodeGroup.LocalFunctions,
+                TypeDeclarationSyntax => NodeGroup.TypeDeclarations,
+                _ => NodeGroup.MemberDeclarations,
+            }))
+        {
+            AddUnnecessaryNodes(semanticModel, group, result, cancellationToken);
+        }
+    }
+
+    public static void AddUnnecessaryNodes(
+        SemanticModel semanticModel,
+        IEnumerable<SyntaxNode> nodesToCheck,
+        ArrayBuilder<SyntaxNode> result,
+        CancellationToken cancellationToken)
+    {
+        using var _2 = PooledDictionary<SyntaxNode, SyntaxAnnotation>.GetInstance(out var nodeToAnnotation);
+        foreach (var node in nodesToCheck)
+            nodeToAnnotation[node] = new SyntaxAnnotation();
+
+        ForkSemanticModelAndCheckNodes(semanticModel, nodesToCheck, nodeToAnnotation, result, cancellationToken);
+    }
+
+    private static void ForkSemanticModelAndCheckNodes(
+        SemanticModel semanticModel,
+        IEnumerable<SyntaxNode> nodesToCheck,
+        PooledDictionary<SyntaxNode, SyntaxAnnotation> nodeToAnnotation,
+        ArrayBuilder<SyntaxNode> result,
+        CancellationToken cancellationToken)
+    {
+        var originalTree = semanticModel.SyntaxTree;
+        var originalRoot = originalTree.GetRoot(cancellationToken);
+
+        var updatedTree = originalTree.WithRootAndOptions(
+            originalRoot.ReplaceNodes(
+                nodesToCheck,
+                (original, current) => WithoutUnsafeModifier(current).WithAdditionalAnnotations(nodeToAnnotation[original])),
+            originalTree.Options);
+        var updateRoot = updatedTree.GetRoot(cancellationToken);
+        var updatedCompilation = semanticModel.Compilation
```

Comment:
Is it easier, possibly more efficient, to do this with `TryGetSpeculativeSemanticModel`?

i.e, get diagnostics of the original declaration. Skip analysis if there are any errors. Then get speculative semantic model with the unsafe keyword removed, and if it still doesn't have errors, then unsafe is unnecessary?

I haven't used speculative semantic model that much, so not sure if it's capable of doing the task here or not.